### PR TITLE
Introduce Json parsing

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Lightweight Json parser based on https://www.json.org/ grammar

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/Json/JsonNode.cs [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+                case Kind.Object:
+                    map = new Dictionary<string, JsonNode>();
+                    break;
+                default:
+                    throw InvalidKind();
+            }
+        }
+
+        public static JsonNode FromString(string json) =>
+            new SyntaxAnalyzer(json).Parse();
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Json/JsonNode.cs:76]: This should be the main entry point. I want the `Parsing` namespace to be hidden from consumers.

File: analyzers/src/SonarAnalyzer.Common/Json/Parsing/LexicalAnalyzer.cs [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+        {
+            Value = null;
+            NextPosition(false);
+            SkipWhiteSpace();               // FIXME: This should be SkipWhitespaceAndComments(), current implementation will throw on comments
+            if (!CanReadCurrentChar)
+            {
+                return Symbol.EOI;
+            }
+            LastStart = CurrentPosition(0);
+            LastEnd = CurrentPosition(1);   // FIXME: Verify that Roslyn needs to be one character behind the end, otherwise, we don't need LastEnd and we don't need int increment in CurrentPosition
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Json/Parsing/LexicalAnalyzer.cs:76]: This should be verified and updated in the PR that will use this in the rules

File: analyzers/src/SonarAnalyzer.Common/Json/Parsing/LexicalAnalyzer.cs [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+        }
+
+        public LinePosition CurrentPosition(int increment) =>
+            new LinePosition(line, position + increment);
+
+        public Symbol NextSymbol()
+        {
+            Value = null;
+            NextPosition(false);
+            SkipWhiteSpace();               // FIXME: This should be SkipWhitespaceAndComments(), current implementation will throw on comments
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Json/Parsing/LexicalAnalyzer.cs:70]: This should be implemented in 2nd PR on the feaure branch

File: analyzers/src/SonarAnalyzer.Common/Json/Parsing/LexicalAnalyzer.cs [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+                {
+                    lines.Add(sb.ToString());
+                    sb.Clear();
+                }
+                else
+                {
+                    sb.Append(c);
+                }
+            }
+            lines.Add(sb.ToString());
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Json/Parsing/LexicalAnalyzer.cs:60]: Using `Span<string>` instead should save some memory, but I'm not sure if we can use it in our codebase

File: analyzers/src/SonarAnalyzer.Common/Json/Parsing/Enums.cs [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        EOI,
+        // Special
+        OpenCurlyBracket,
+        CloseCurlyBracket,
+        OpenSquareBracket,
+        CloseSquareBracket,
+        Comma,
+        Colon,
+        // Terms
+        Value   // String, Number (Integer, Decimal, Double), Null, True, False
```
Comment (Corniel) [inline analyzers/src/SonarAnalyzer.Common/Json/Parsing/Enums.cs:41]: Is this intended as documentation? (If so, why not use `/// <summary>` or as TODO? (in that case, mark it as such).

File: analyzers/src/SonarAnalyzer.Common/Json/Parsing/LexicalAnalyzer.cs (hunk 4 above)
Comment (Corniel) [reply analyzers/src/SonarAnalyzer.Common/Json/Parsing/LexicalAnalyzer.cs:60]: What are the objections? (Just curious).

Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Json/Parsing/LexicalAnalyzer.cs:60]: There are no objections, as it is a clear winner. But we target `net46` as a technical limitation.

File: analyzers/src/SonarAnalyzer.Common/Json/Parsing/Enums.cs (hunk 5 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Json/Parsing/Enums.cs:41]: This is just an explanatory comment for me

File: analyzers/src/SonarAnalyzer.Common/Json/Parsing/LexicalAnalyzer.cs (hunk 4 above)
Comment (Corniel) [reply analyzers/src/SonarAnalyzer.Common/Json/Parsing/LexicalAnalyzer.cs:60]: Oops, fair point.

**Review by andrei-epure-sonarsource — APPROVED** [review]
LGTM

File: analyzers/src/SonarAnalyzer.Common/Json/JsonException.cs [hunk 6]
```diff
@@ -21,13 +21,14 @@
...
 using System.Runtime.Serialization;
+using Microsoft.CodeAnalysis.Text;
 
 namespace SonarAnalyzer.Json
 {
     [Serializable]
     public sealed class JsonException : Exception
     {
-        public JsonException(string message) : base(message) { }
+        public JsonException(string message, LinePosition position) : base(message + $" at line {position.Line + 1} position {position.Character + 1}") { }
```
[the remark this answers is not in the document]
Comment (Corniel) [reply analyzers/src/SonarAnalyzer.Common/Json/JsonException.cs:31]: I would argue that it is not a JsonException, but a ParseError, or ParseException. but that is also nitpicking.

**Comment by andrei-epure-sonarsource** [comment]
This has been merged to the feature branch so should not be in Validate Peach column because it won't be deployed.

**Pull request merged**